### PR TITLE
[Console] improve exception message when required argument is added after an optional one

### LINKED ISSUE #40302: Exceptions thrown in InputDefinition should specify the command and the argument
**Symfony version(s) affected**: 5.x

**Description**

The exceptions thrown in InputDefinition don't tell you where the problem is.

**How to reproduce**

I wrote a command and messed up the @param in the docblock. I got an exception "Cannot add a required argument after an optional one." But which command? which param? I could have installed lots of packages at once which all provided commands to something such as Drush, and I would have no way of knowing which command caused the problem.

**Possible Solution**

Add the command name or the command method name and the argument to the exception message.

**Additional context**

### PR DESCRIPTION
**Pull request by ghost**
| Q             | A
| ------------- | ---
| Branch?       | 5.x 
| Bug fix?      | no
| New feature?  | wouldn't say so, rather DX improvement 
| Deprecations? | no 
| Tickets       | Fix #40302 
| License       | MIT
| Doc PR        | n/a

Hello, this is my first contribution to Symfony Framework. It's time to pull my weight.
About the issue:
I did improve an error message to include passed argument's name and the latest optional argument's name.
Screenshot at 2021-02-26 23-08-10
An author also mentioned "But which command?", however this is shown twice as we can see on the screenshot above so I skipped that.

### CODE REVIEW

File: src/Symfony/Component/Console/Input/InputDefinition.php [hunk 1]
```diff
@@ -30,8 +30,8 @@ class InputDefinition
 {
     private $arguments;
     private $requiredCount;
-    private $hasAnArrayArgument = false;
-    private $hasOptional;
+    private $lastArrayArgument;
+    private $lastOptional;
```
Comment (chalasr) [inline src/Symfony/Component/Console/Input/InputDefinition.php]: `lastOptionalArgument`?

File: src/Symfony/Component/Console/Tests/Input/InputDefinitionTest.php [hunk 2]
```diff
@@ -101,7 +101,7 @@ public function testArgumentsMustHaveDifferentNames()
     public function testArrayArgumentHasToBeLast()
     {
         $this->expectException(\LogicException::class);
-        $this->expectExceptionMessage('Cannot add an argument after an array argument.');
+        $this->expectExceptionMessage(sprintf('Cannot add a required argument "%s" after an optional one "%s".', 'anotherbar', 'fooarray'));
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Input/InputDefinitionTest.php]: sprintf should not be called with hardcoded values

**Review by chalasr — CHANGES_REQUESTED** [review]
Nice first contribution :)

File: src/Symfony/Component/Console/Input/InputDefinition.php [hunk 3]
```diff
@@ -72,8 +72,8 @@ public function setArguments(array $arguments = [])
     {
         $this->arguments = [];
         $this->requiredCount = 0;
-        $this->hasOptional = false;
-        $this->hasAnArrayArgument = false;
+        $this->lastOptional = null;
```
Comment (OskarStark) [inline src/Symfony/Component/Console/Input/InputDefinition.php]: Please keep the same order as the properties

Comment (marcin-bulkiewicz) [reply src/Symfony/Component/Console/Input/InputDefinition.php]: It was like that before. I just bulk changed this field using my IDE feature.

Comment (ghost) [reply src/Symfony/Component/Console/Input/InputDefinition.php]: It was like that before. I've just bulk changed this field's name with my IDE feature.

**Comment by OskarStark** [comment]
It looks like your commiter email is not associated with your github account, can you please check? Thanks

File: src/Symfony/Component/Console/Input/InputDefinition.php [hunk 4]
```diff
@@ -100,22 +100,22 @@ public function addArgument(InputArgument $argument)
             throw new LogicException(sprintf('An argument with name "%s" already exists.', $argument->getName()));
         }
 
-        if ($this->hasAnArrayArgument) {
-            throw new LogicException('Cannot add an argument after an array argument.');
+        if (null !== $this->lastArrayArgument) {
+            throw new LogicException(sprintf('Cannot add a required argument "%s" after an optional one "%s".', $argument->getName(), $this->lastArrayArgument->getName()));
```
Comment (stof) [inline src/Symfony/Component/Console/Input/InputDefinition.php]: this is the wrong error message. This error happens when adding an argument after an array argument, not when adding a required argument after an optional one (which is the next check).
I suspect a bad copy-paste.

File: src/Symfony/Component/Console/Tests/Input/InputDefinitionTest.php [hunk 5]
```diff
@@ -101,7 +101,7 @@ public function testArgumentsMustHaveDifferentNames()
     public function testArrayArgumentHasToBeLast()
     {
         $this->expectException(\LogicException::class);
-        $this->expectExceptionMessage('Cannot add an argument after an array argument.');
+        $this->expectExceptionMessage('Cannot add a required argument "anotherbar" after an optional one "fooarray".');
```
Comment (stof) [inline src/Symfony/Component/Console/Tests/Input/InputDefinitionTest.php]: this test is expecting the wrong error message too.

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
> It looks like your commiter email is not associated with your github account, can you please check? Thanks

Where did you spot this?

File: src/Symfony/Component/Console/Input/InputDefinition.php (hunk 4 above)
Comment (ghost) [reply src/Symfony/Component/Console/Input/InputDefinition.php]: Yeah, I think it was a copy past.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @marbulk.

**Pull request merged**